File: frontend/src/types/worker.ts
/**
 * Worker Types
 */

export interface GPUInfo {
  index: number;
  name: string;
  memory_total: number;
  memory_used: number;
  memory_free: number;
  utilization: number;
  temperature?: number;
}

export interface CPUInfo {
  percent: number;
  count: number;
  freq_mhz: number;
}

export interface MemoryInfo {
  total: number;
  used: number;
  free: number;
  percent: number;
}

export interface DiskInfo {
  total: number;
  used: number;
  free: number;
  percent: number;
}

export interface CapabilitiesInfo {
  os_type: string;
  gpu_type: string;
  docker: boolean;
  ollama: boolean;
  ollama_running: boolean;
  mlx?: boolean;
  llama_cpp?: boolean;
}

export interface SystemInfo {
  cpu?: CPUInfo;
  memory?: MemoryInfo;
  disk?: DiskInfo;
  os_type?: string;
  gpu_type?: string;
  capabilities?: CapabilitiesInfo;
}

export type ConnectionType = "direct" | "tailscale";
export type OSType = "linux" | "darwin" | "windows";
export type GPUType = "nvidia" | "apple_silicon" | "amd" | "none";

export interface Worker {
  id: number;
  name: string;
  address: string;
  status: "online" | "offline" | "error";
  connection_type: ConnectionType;
  tailscale_ip?: string;
  headscale_node_id?: number;
  effective_address?: string;
  description?: string;
  labels?: Record<string, string>;
  gpu_info?: GPUInfo[];
  system_info?: SystemInfo;
  os_type: OSType;
  gpu_type: GPUType;
  capabilities?: CapabilitiesInfo;
  available_backends?: string[];
  created_at: string;
  updated_at: string;
  last_heartbeat?: string;
  deployment_count: number;
}

export interface WorkerCreate {
  name: string;
  address: string;
  description?: string;
  labels?: Record<string, string>;
  connection_type?: ConnectionType;
  tailscale_ip?: string;
  headscale_node_id?: number;
}

export interface RegistrationToken {
  id: number;
  token: string;
  name: string;
  is_used: boolean;
  used_by_worker_id?: number;
  created_at: string;
  expires_at: string;
  used_at?: string;
  is_valid: boolean;
  docker_command?: string;
}

export interface RegistrationTokenCreate {
  name: string;
  expires_in_hours?: number;
}

export interface LocalWorkerSpawnResponse {
  success: boolean;
  message: string;
  worker_name: string;
  container_id?: string;
  backend_url: string;
}

File: backend/app/schemas/worker.py
"""Worker Pydantic schemas"""

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.worker import ConnectionType, WorkerStatus


class GPUInfo(BaseModel):
    """GPU information schema"""

    index: int
    name: str
    memory_total: int  # bytes
    memory_used: int = 0
    memory_free: int = 0
    utilization: int = 0  # percentage
    temperature: int = 0  # Celsius


class CPUInfo(BaseModel):
    """CPU information schema"""

    percent: float = 0  # CPU usage percentage
    count: int = 0  # Number of CPU cores
    freq_mhz: float = 0  # CPU frequency in MHz


class MemoryInfo(BaseModel):
    """Memory information schema"""

    total: int = 0  # bytes
    used: int = 0  # bytes
    free: int = 0  # bytes
    percent: float = 0  # percentage


class DiskInfo(BaseModel):
    """Disk information schema"""

    total: int = 0  # bytes
    used: int = 0  # bytes
    free: int = 0  # bytes
    percent: float = 0  # percentage


class SystemInfo(BaseModel):
    """System information schema (CPU, Memory, Disk)"""

    cpu: CPUInfo | None = None
    memory: MemoryInfo | None = None
    disk: DiskInfo | None = None


class WorkerBase(BaseModel):
    """Base worker schema"""

    name: str = Field(..., min_length=1, max_length=255)
    address: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    labels: dict | None = None
    connection_type: str = Field(
        default=ConnectionType.DIRECT.value,
        description="Connection type: direct or tailscale",
    )


class WorkerCreate(WorkerBase):
    """Schema for creating a worker"""

    gpu_info: list[GPUInfo] | None = None
    system_info: SystemInfo | None = None
    tailscale_ip: str | None = Field(None, description="IP address in Tailscale network")
    headscale_node_id: int | None = Field(None, description="Node ID in Headscale")


class WorkerUpdate(BaseModel):
    """Schema for updating a worker"""

    name: str | None = Field(None, min_length=1, max_length=255)
    address: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    labels: dict | None = None
    status: WorkerStatus | None = None
    gpu_info: list[GPUInfo] | None = None
    system_info: SystemInfo | None = None
    connection_type: str | None = Field(None, description="Connection type: direct or tailscale")
    tailscale_ip: str | None = Field(None, description="IP address in Tailscale network")
    headscale_node_id: int | None = Field(None, description="Node ID in Headscale")


class WorkerResponse(WorkerBase):
    """Schema for worker response"""

    id: int
    status: str
    gpu_info: list[dict] | None = None
    system_info: dict | None = None
    tailscale_ip: str | None = None
    headscale_node_id: int | None = None
    effective_address: str | None = None  # The actual address to use for connections
    created_at: datetime
    updated_at: datetime
    last_heartbeat: datetime | None = None
    deployment_count: int = 0

    class Config:
        from_attributes = True


class WorkerListResponse(BaseModel):
    """Schema for worker list response"""

    items: list[WorkerResponse]
    total: int


class WorkerHeartbeat(BaseModel):
    """Schema for worker heartbeat"""

    worker_id: int
    gpu_info: list[GPUInfo] | None = None
    system_info: SystemInfo | None = None
    status: WorkerStatus = WorkerStatus.ONLINE
